quodlibet/util/matcher.py

```diff
@@ -0,0 +1,382 @@
+
+        # normalize
+        weight_sum = sum(attr_to_weight.values())
+
+        self._attr_with_weight = []
+        for attr, weight in attr_to_weight.items():
+            if weight <= 0:
+                raise ValueError("weights <= 0 are not allowed")
+
+            self._attr_with_weight.append((attr, weight / weight_sum))
+
+        # from largest to smallest weight
+        self._attr_with_weight.sort(key=lambda i: i[1], reverse=True)
+
+        weight_left = 1.0
+        self._weight_left = []
+        for _, weight in self._attr_with_weight:
+            self._weight_left.append(weight_left)
+            weight_left -= weight
+
+    def get_indices(self, a_items: List[T], b_items: List[T]) -> List[Optional[int]]:
+        """
+        :return: the indices of b ordered so that they match elements in a.
+
+        Size of a_items and b_items can differ. If there are more a_items than
+        there are b_items, None is used if no match could be assigned to an a.
+        The returned list always has the size of a_items.
+
+        In terms of performance, it's preferable to supply the smaller list as a_items.
+        """
+        if not b_items:
+            return [None for _ in a_items]
+
+        self._b_items = b_items
+        b_size = len(b_items)
+
+        self._b_idx_to_a_match_data = [None for _ in range(b_size)]
+        self._a_idx_to_match_data = []
+
+        for a_idx, a in enumerate(a_items):
+            match_data = _MatchData(a_idx, a, b_size)
+            self._a_idx_to_match_data.append(match_data)
+
+            self._measure_similarity_to_find_best_b_match(match_data)
+
+            if self._is_below_minimum_similarity(match_data):
+                match_data.best_b_idx = None
+            else:
+                self._handle_conflicts_if_any(match_data)
+
+        result = [md.best_b_idx for md in self._a_idx_to_match_data]
+
+        if self.should_store_similarity_matrix:
+            matrix = [md.b_idx_to_similarity for md in self._a_idx_to_match_data]
+            self.similarity_matrix = matrix
+
+        # cleanup
+        del self._b_items
+        del self._b_idx_to_a_match_data
+        del self._a_idx_to_match_data
+
+        return result
+
+    def _is_below_minimum_similarity(self, match_data):
+        cont_attr_idx = match_data.continue_attr_index
+        similarity = match_data.best_b_similarity
+
+        # If this hasn't gone through every attribute and weight yet, we normalize the
+        # similarity, so it's in the range of [0, 1]. Without this, minimum_similarity
+        # would be somewhat unpredictable.
+        #
+        # For example, a b_item might be undefeatable with a similarity of 0.52. If we
+        # wouldn't normalize it, it would not pass a minimum of 0.6, even if it could
+        # easily achieve that with the remaining weights.
+        if cont_attr_idx < len(self._attr_with_weight):
+            cur_total_weight = 1 - self._weight_left[cont_attr_idx]
```

Comment:
How do we know `self._weight_left[cont_attr_idx]` isn't 1?

Reply:
The `_measure_similarity_to_find_best_b_match` always goes through at least one attribute, so `cont_attr_idx` can never be 0 and per definition this can never be 1:

https://github.com/quodlibet/quodlibet/blob/b838ebcae22b0b2c63944f7663bf8e07063849cd/quodlibet/util/matcher.py#L184-L188

(before that section we have checked that no weight is 0)

Reply:
OK cool - maybe add a comment here to say we _know_ it's not zero, for the next reader :+1: